app/api/v2/managers/operation_api_manager.py

```diff
@@ -20,3 +49,51 @@ async def get_operation_object(self, operation_id: str, access: dict):
         if operation.match(access):
             return operation
         raise JsonHttpForbidden(f'Cannot view operation due to insufficient permissions: {operation_id}')
+
+    async def setup_operation(self, data: dict, access: BaseWorld.Access):
+        """Applies default settings to an operation if data is missing."""
+        planner_id = data.pop('planner', {}).get('id', '')
+        data['planner'] = await self._construct_and_dump_planner(planner_id)
+        adversary_id = data.pop('adversary', {}).get('adversary_id', '')
+        data['adversary'] = await self._construct_and_dump_adversary(adversary_id)
+        fact_source_id = data.pop('source', {}).get('id', '')
+        data['source'] = await self._construct_and_dump_source(fact_source_id)
+        operation = OperationSchema().load(data)
+        await operation.update_operation_agents(self.services)
+        allowed = self._get_allowed_from_access(access)
+        operation.set_operation_access(allowed)
+        operation.set_start_details()
+        return operation
+
+    async def _construct_and_dump_planner(self, planner_id: str):
+        planner = await self.services['data_svc'].locate('planners', match=dict(id=planner_id))
+        if not planner:
+            planner = await self.services['data_svc'].locate('planners', match=dict(name='atomic'))
+        return PlannerSchema().dump(planner[0])
+
+    async def _construct_and_dump_adversary(self, adversary_id: str):
+        adv = await self.services['data_svc'].locate('adversaries', match=dict(adversary_id=adversary_id))
+        if not adv:
+            adv = Adversary.load(dict(adversary_id='ad-hoc', name='ad-hoc', description='an empty adversary profile',
+                                      atomic_ordering=[]))
+        else:
+            adv = adv[0]
+        return AdversarySchema().dump(adv)
+
+    async def _construct_and_dump_source(self, source_id: str):
+        source = await self.services['data_svc'].locate('sources', match=dict(id=source_id))
+        if not source:
+            source = (await self.services['data_svc'].locate('sources', match=dict(name='basic')))
+        return SourceSchema().dump(source[0])
+
+    async def validate_operation_state(self, data: dict, existing: Operation = None):
+        if not existing:
+            if data.get('state') in Operation.get_finished_states():
+                raise JsonHttpBadRequest('Cannot create a finished operation.')
+            elif data.get('state') not in Operation.get_states():
+                raise JsonHttpBadRequest('state must be one of {}'.format(Operation.get_states()))
+        else:
+            if await existing.is_finished() and data.get('state') not in Operation.get_finished_states():
```

Comment:
I don't think this will catch the edge case where the operation is in the "finished" state, and we try to go to "cleanup", will it?

Reply:
I think if the operation is in the "finished" or "out of time" status, maybe we don't allow any changes at all.

Reply:
Actually I guess it doesn't matter since the operation will be finished anyway

Reply:
operation.is_finished() checks for all the finished statuses (finished, clean up, out of time) but I guess it doesn't make too much sense to be able to change finished status. I'll update this